sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GeneratedColumn.scala

```
@@ -0,0 +1,160 @@
+          s"Failed to execute $statementType command because the column $fieldName has " +
+            s"generation expression $expressionStr which fails to parse as a valid expression:" +
+            s"\n${ex.getMessage}")
+    }
+    // Analyze the parse result
+    // Generated column can't reference itself
+    val relation = new LocalRelation(StructType(schema.filterNot(_.name == fieldName)).toAttributes)
+    val plan = try {
+      val analyzer: Analyzer = GeneratedColumnAnalyzer
+      val analyzed = analyzer.execute(Project(Seq(Alias(parsed, fieldName)()), relation))
+      analyzer.checkAnalysis(analyzed)
+      analyzed
+    } catch {
+      case ex: AnalysisException =>
+        // Improve error message if possible
+        if (ex.getErrorClass == "UNRESOLVED_COLUMN.WITH_SUGGESTION") {
+          ex.messageParameters.get("objectName").filter(_ == toSQLId(fieldName)).foreach { _ =>
+            // Generation expression references itself
+            throw new AnalysisException(
+              errorClass = "UNSUPPORTED_EXPRESSION_GENERATED_COLUMN",
+              messageParameters = Map(
+                "fieldName" -> fieldName,
+                "expressionStr" -> expressionStr,
+                "reason" -> "generation expression cannot reference itself",
+                "errorMessage" -> ex.getMessage))
+          }
+        }
+        if (ex.getErrorClass == "UNRESOLVED_ROUTINE") {
+          // Cannot resolve function using built-in catalog
+          ex.messageParameters.get("routineName").foreach { fnName =>
+            throw new AnalysisException(
+              errorClass = "UNSUPPORTED_EXPRESSION_GENERATED_COLUMN",
+              messageParameters = Map(
+                "fieldName" -> fieldName,
+                "expressionStr" -> expressionStr,
+                "reason" -> s"failed to resolve $fnName to a built-in function",
+                "errorMessage" -> ex.getMessage))
+          }
+        }
+        throw new AnalysisException(
```

shall we just rethrow the original exception? We have error context which will point to the exact place in the SQL statement that trigger the error, and we don't need to highlight here that this is a generated expression.

---

+1